tests/e2e/050_nat_test.go

```diff
@@ -0,0 +1,516 @@
+		HostIfName:  linuxTap2Hostname,
+		Link: &linux_interfaces.Interface_Tap{
+			Tap: &linux_interfaces.TapLink{
+				VppTapIfName: vppTap2Name,
+			},
+		},
+		Namespace: &linux_namespace.NetNamespace{
+			Type:      linux_namespace.NetNamespace_MICROSERVICE,
+			Reference: msNamePrefix + ms2Name,
+		},
+	}
+
+	ms2DefaultRoute := &linux_l3.Route{
+		OutgoingInterface: linuxTap2Name,
+		Scope:             linux_l3.Route_GLOBAL,
+		DstNetwork:        "0.0.0.0/0",
+		GwAddr:            vppTap2IP,
+	}
+
+	natGlobal := &vpp_nat.Nat44Global{
+		Forwarding: true,
+		NatInterfaces: []*vpp_nat.Nat44Global_Interface{
+			{
+				Name:     vppTap1Name,
+				IsInside: false,
+			},
+			{
+				Name:     vppTap2Name,
+				IsInside: true,
+			},
+		},
+	}
+
+	showTCPSvc := fmt.Sprintf("tcp local %s:%d external %s:%d vrf 0  out2in-only",
+		linuxTap2IP, tcpSvcLocalPort, tcpSvcExtIP, tcpSvcExtPort)
+	tcpSvc := &vpp_nat.DNat44{
+		Label: tcpSvcLabel,
+		StMappings: []*vpp_nat.DNat44_StaticMapping{
+			{
+				ExternalIp:   tcpSvcExtIP,
+				ExternalPort: tcpSvcExtPort,
+				Protocol:     vpp_nat.DNat44_TCP,
+				LocalIps: []*vpp_nat.DNat44_StaticMapping_LocalIP{
+					{
+						LocalIp:   linuxTap2IP,
+						LocalPort: tcpSvcLocalPort,
+					},
+				},
+			},
+		},
+	}
+
+	showUDPSvc := fmt.Sprintf("udp local %s:%d external %s:%d vrf 0  out2in-only",
+		linuxTap2IP, udpSvcLocalPort, udpSvcExtIP, udpSvcExtPort)
+	udpSvc := &vpp_nat.DNat44{
+		Label: udpSvcLabel,
+		StMappings: []*vpp_nat.DNat44_StaticMapping{
+			{
+				ExternalIp:   udpSvcExtIP,
+				ExternalPort: udpSvcExtPort,
+				Protocol:     vpp_nat.DNat44_UDP,
+				LocalIps: []*vpp_nat.DNat44_StaticMapping_LocalIP{
+					{
+						LocalIp:   linuxTap2IP,
+						LocalPort: udpSvcLocalPort,
+					},
+				},
+			},
+		},
+	}
+
+	// without proxy ARP, the client running from microservice1 (public net),
+	// will not get any response for ARP requests concerning service external IPs
+	// from VPP
+	svcExtIPsProxyArp := &vpp_l3.ProxyARP{
+		Interfaces: []*vpp_l3.ProxyARP_Interface{
+			{
+				Name: vppTap1Name,
+			},
+		},
+		Ranges: []*vpp_l3.ProxyARP_Range{
+			{
+				FirstIpAddr: tcpSvcExtIP,
+				LastIpAddr:  tcpSvcExtIP,
+			},
+			{
+				FirstIpAddr: udpSvcExtIP,
+				LastIpAddr:  udpSvcExtIP,
+			},
+		},
+	}
+
+	checkConfig := func(isConfigured bool) {
+		substringMatcher := func(substring string, negative bool) types.GomegaMatcher {
+			if negative {
+				return Not(ContainSubstring(substring))
+			} else {
+				return ContainSubstring(substring)
+			}
+		}
+		stdout, err := ctx.execVppctl("show", "nat44", "static", "mappings")
+		Expect(err).ToNot(HaveOccurred(), "Running `vppctl show nat44 addresses` failed")
+		Expect(stdout).To(substringMatcher(showTCPSvc, !isConfigured),
+			"Unexpected TCP static mapping configuration",
+		)
+		Expect(stdout).To(substringMatcher(showUDPSvc, !isConfigured),
+			"Unexpected UDP static mapping configuration",
+		)
+	}
+
+	checkConn := func(shouldSucceed bool) {
```

Comment:
Same applies here. The tests should try to be readable with the least indirection. Also, if this test fails it will report line number inside this helper, you would need to call `ExpectWithOffset(1, ...) to get proper line nubmer.

However you could simply define helper functions for each connect type: 

```go
connectTCP := func() error {
	return ctx.testConnection(ms1Name, ms2Name, tcpSvcExtIP, linuxTap2IP,
		tcpSvcExtPort, tcpSvcLocalPort, false, tapv2InputNode)
}
connectUDP := // ...
```

then the assertions will be more verbose, but more understandable to users without need to go to the helper function to understand what `checkConn` really does:

```go
checkConn(true)
// vs.
Expect(connectTCP()).To(Succeed())
Expect(connectUDP()).To(Succeed())
```